keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,369 @@
+       // +optional
+       PVCList []PersistentVolumeClaim
+
+      Ready *bool
+  
+      // Last error encountered during group creation
+      Error *VolumeGroupError
+}
+
+// Describes an error encountered on the group
+type VolumeGroupError struct {
+    	// time is the timestamp when the error was encountered.
+    	// +optional
+    	Time *metav1.Time
+ 
+    	// message details the encountered error
+    	// +optional
+    	Message *string
+}
+
+// GroupSnapshot is a user's request for taking a group snapshot. 
+type GroupSnapshot struct {
+        metav1.TypeMeta `json:",inline"`
+        // Standard object's metadata.
+        // +optional
+        metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+        // Spec defines the desired characteristics of a group snapshot requested by a user.
+        Spec GroupSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+        // Status represents the latest observed state of the group snapshot
+        // +optional
+        Status *GroupSnapshotStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// GroupSnapshotSpec describes the common attributes of a group snapshot
+type GroupSnapshotSpec struct {
+        // Source has the information about where the group snapshot is created from.
+        // Supported Kind is VolumeGroup
+        // +optional
+        Source *TypedLocalObjectReference `json:"source" protobuf:"bytes,1,opt,name=source"`
+
+        // GroupSnapshotName is the name of the group snapshot
+        // +optional
+        GroupSnapshotName string `json:"groupSnapshotName" protobuf:"bytes,2,opt,name=groupSnapshotName"`
+}
+
+Type GroupSnapshotStatus struct {
+
+        // ReadyToUse becomes true when ReadyToUse on all individual snapshots become true
+        // +optional
+        ReadyToUse *bool
+
+        // List of volume snapshots
+        SnapshotList []VolumeSnapshot
+}
+
+type PersistentVolumeClaimSpec struct {
+	......
+        VolumeGroupId *string
+	......
+}
+
+
+type VolumeSnapshotSpec struct{
+	......
+        GroupSnapshotId *string
+	......
+}
+```
+
+### Example Yaml Files
+
+#### Volume Group Snapshot
+
+Example yaml files to define a VolumeGroupClass and VolumeGroup are in the following.
+
+A VolumeGroupClass that supports groupSnapshot:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupClass
+metadata:
+  name: volumeGroupClass1
+spec:
+  parameters:
+     …...
+  groupSnapshot: true
+```
+
+A VolumeGroup belongs to this VolumeGroupClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroup
+metadata:
+  Name: volumeGroup1
+spec:
+  volumeGroupClassName: volumeGroupClass1
+  selector:
+    matchLabels:
+      app: my-app
+```
+
+A GroupSnapshot taken from the VolumeGroup:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: GroupSnapshot
+metadata:
+  name: my-group-snapshot
+spec:
+  sourceGroupName: volumeGroup1
```

Comment:
I don't find this in the spec definition

Reply:
This is "Source *TypedLocalObjectReference".  Name needs to be fixed.

Reply:
Updated.